code/lib/cli/src/automigrate/fixes/missing-storybook-dependencies.ts

```diff
@@ -0,0 +1,156 @@
+import chalk from 'chalk';
+import { readFile } from 'node:fs/promises';
+import { dedent } from 'ts-dedent';
+
+import type { Fix } from '../types';
+import { getStorybookVersionSpecifier } from '../../helpers';
+import type { InstallationMetadata, JsPackageManager } from '@storybook/core/common';
+
+const logger = console;
+
+type PackageUsage = Record<string, string[]>;
+
+interface MissingStorybookDependenciesOptions {
+  packageUsage: PackageUsage;
+}
+
+const consolidatedPackages = [
+  '@storybook/channels',
+  '@storybook/client-logger',
+  '@storybook/core-common',
+  '@storybook/core-events',
+  '@storybook/csf-tools',
+  '@storybook/docs-tools',
+  '@storybook/node-logger',
+  '@storybook/preview-api',
+  '@storybook/router',
+  '@storybook/telemetry',
+  '@storybook/theming',
+  '@storybook/types',
+  '@storybook/manager-api',
+  '@storybook/manager',
+  '@storybook/preview',
+  '@storybook/core-server',
+  '@storybook/builder-manager',
+  '@storybook/components',
+];
+
+async function checkInstallations(
+  packageManager: JsPackageManager,
+  packages: string[]
+): Promise<InstallationMetadata['dependencies']> {
+  let result: Record<string, any> = {};
+
+  // go through each package and get installation info at depth 0 to make sure
+  // the dependency is directly installed, else they could come from other dependencies
+  const promises = packages.map((pkg) => packageManager.findInstallations([pkg], { depth: 0 }));
+
+  const analyses = await Promise.all(promises);
```

Comment:
This unfortunately adds several seconds in the check, in contrast with just passing the entire array as a pattern. The reason I do this is because it's possible that in yarn, if you pass a package that does not exist, an error will be thrown. As a result, you get `undefined` and then lose the information regarding all packages which is not great. Iterating through the packages instead fixes that, at the cost of slower times because then it does multiple requests instead

Reply:
How about we change it so the slow implementation is only in yarn1?

So we keep passing the array to `packageManager.findInstallations` but then inside the yarn1 implementation we do the `Promise.all` and split it over multiple processes?

That way all the other package managers perf are unaffected.

Reply:
The Promise.all solution was for both yarn 1 and 2! I didn’t test on npm and pnpm to know whether it would work there as well

Reply:
I see. 👍